lib/grammers-client/src/client/files.rs

```diff
@@ -325,6 +343,60 @@ impl Client {
         // files, so it's fine to unwrap.
         let name = path.file_name().unwrap().to_string_lossy().to_string();
 
-        self.upload_stream(&mut file, size, name).await
+        self.upload_stream(file, size, name).await
+    }
+}
+
+struct PartStream<S: AsyncRead + Unpin + Send + 'static> {
+    stream: Mutex<S>,
+    total_parts: i32,
+    current_part: AtomicI32,
+}
+
+impl<S: AsyncRead + Unpin + Send> PartStream<S> {
+    fn new(stream: S, size: usize) -> Self {
+        let total_parts = ((size + MAX_CHUNK_SIZE as usize - 1) / MAX_CHUNK_SIZE as usize) as i32;
+        Self {
+            stream: Mutex::new(stream),
+            total_parts,
+            current_part: AtomicI32::new(0),
+        }
+    }
+
+    fn total_parts(&self) -> i32 {
+        self.total_parts
+    }
+
+    async fn next_part(&self) -> Result<Option<(i32, Vec<u8>)>, io::Error> {
+        let mut stream = self.stream.lock().await;
```

Comment:
Maybe we should use our custom mutex here so that they can be debugged too.

Reply:
Done